src/js/client/handlebars.js

```diff
@@ -30,63 +31,56 @@
         }
     };
 
-    gpii.templates.hb.client.render = function (that, key, context) {
-        // TODO:  Convert to "that-ism" where we use locate() instead of $(selector)
-        // If a template exists, load that.  Otherwise, try to load the partial.
-        var element = $("#partial-" + key).length ? $("#partial-" + key) : $("#template-" + key);
+    gpii.templates.hb.client.render = function (that, templateName, context) {
+        var templateType = that.templates.partials[templateName] ? "partials" : that.templates.pages[templateName] ? "pages" : that.templates.layouts[templateName] ? "layouts" : null;
+        if (!templateType) {
+            fluid.fail("Can't find template '" + templateName + "'...");
+        }
 
         // Cache each compiled template the first time we use it...
-        if (that.compiled[key]) {
-            return that.compiled[key](context);
-        }
-        else {
-            if (!element || !element.html()) {
-                fluid.fail("Template '" + key + "' does not have any content. Skipping");
-                return;
-            }
+        if (!that.compiled[templateType][templateName]) {
+            var compiledTemplate = Handlebars.compile(that.templates[templateType][templateName]);
 
-            var template = Handlebars.compile(element.html());
-            that.compiled[key] = template;
-            return template(context);
+            that.compiled[templateType][templateName] = compiledTemplate;
         }
+
+        return that.compiled[templateType][templateName](context);
     };
 
     gpii.templates.hb.client.passthrough = function (that, element, key, context, manipulator) {
-        // TODO: Confirm whether that.function syntax works here
         element[manipulator](gpii.templates.hb.client.render(that, key, context));
     };
 
     ["after", "append", "before", "prepend", "replaceWith", "html"].forEach(function (manipulator) {
-        // TODO: Confirm whether that.function syntax works here
         gpii.templates.hb.client[manipulator] = function (that, element, key, context) {
             gpii.templates.hb.client.passthrough(that, element, key, context, manipulator);
         };
     });
 
-    gpii.templates.hb.client.appendToBody = function (that, data) {
-        // TODO:  Replace this with a {that} reference?
-        $("body").append(data);
+    gpii.templates.hb.client.cacheTemplates = function (that, data) {
+        data = typeof data === "string" ? JSON.parse(data) : data;
```

Comment:
Another example of the awkward logic that is duplicated in gpii.templates.hb.client.templateFormControl.jsonOrString . It will be safest to just ensure that exactly one of either you or jQuery takes care of parsing a JSON payload exactly once (the latter option via supplying dataType: "json" to the request, the former by calling JSON.parse)